eth/stagedsync/stage_tevm.go

```diff
@@ -0,0 +1,344 @@
+package stagedsync
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"time"
+
+	"github.com/c2h5oh/datasize"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/ledgerwatch/erigon/common"
+	"github.com/ledgerwatch/erigon/common/dbutils"
+	"github.com/ledgerwatch/erigon/eth/stagedsync/stages"
+	"github.com/ledgerwatch/erigon/ethdb"
+	"github.com/ledgerwatch/erigon/log"
+	"github.com/ledgerwatch/erigon/metrics"
+	"github.com/ledgerwatch/erigon/params"
+)
+
+var stageTranspileGauge = metrics.NewRegisteredGauge("stage/tevm", nil)
+
+type TranspileCfg struct {
+	db            ethdb.RwKV
+	batchSize     datasize.ByteSize
+	readerBuilder StateReaderBuilder
+	writerBuilder StateWriterBuilder
+	chainConfig   *params.ChainConfig
+}
+
+func StageTranspileCfg(
+	kv ethdb.RwKV,
+	batchSize datasize.ByteSize,
+	readerBuilder StateReaderBuilder,
+	writerBuilder StateWriterBuilder,
+	chainConfig *params.ChainConfig,
+) TranspileCfg {
+	return TranspileCfg{
+		db:            kv,
+		batchSize:     batchSize,
+		readerBuilder: readerBuilder,
+		writerBuilder: writerBuilder,
+		chainConfig:   chainConfig,
+	}
+}
+
+type contract struct {
+	hash []byte
+	code []byte
+}
+
+func transpileBatch(logPrefix string, s *StageState, fromBlock uint64, toBlock uint64, tx ethdb.RwTx, batch ethdb.DbWithPendingMutations, cfg TranspileCfg, useExternalTx bool, quitCh <-chan struct{}) error {
+	done := make(chan struct{})
+	inContracts := make(chan contract, 8192)
+	outContracts := make(chan contract, 8192)
+
+	readWG := new(errgroup.Group)
+	numWorkers := runtime.NumCPU()
```

Comment:
I would first start with 1 go-routine to see how fast/slow this is, and then parallelise when needed